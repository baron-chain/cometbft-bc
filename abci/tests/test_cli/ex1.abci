package tests

import (
    "testing"
    "context"
    "time"
    "encoding/hex"

    "github.com/stretchr/testify/require"
    "github.com/stretchr/testify/assert"

    "github.com/baron-chain/cometbft-bc/abci/types"
)

const (
    testTimeout = 30 * time.Second
)

// BCTestCommand represents a Baron Chain test command
type BCTestCommand struct {
    name string
    exec func(*testing.T, types.ABCIClient)
}

func TestBaronChainSequence(t *testing.T) {
    ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
    defer cancel()

    env := newBCTestEnv(t)
    defer env.stop()
    env.start()

    testSequences := []struct {
        name     string
        commands []BCTestCommand
    }{
        {
            name: "Basic Baron Chain Operations",
            commands: []BCTestCommand{
                {
                    name: "Echo Verification",
                    exec: func(t *testing.T, client types.ABCIClient) {
                        res, err := client.EchoSync("baron-chain")
                        require.NoError(t, err, "echo failed")
                        assert.Equal(t, "baron-chain", res.Message)
                    },
                },
                {
                    name: "Info Verification",
                    exec: func(t *testing.T, client types.ABCIClient) {
                        res, err := client.InfoSync(types.RequestInfo{Version: "1.0"})
                        require.NoError(t, err, "info check failed")
                        assert.NotNil(t, res, "empty info response")
                    },
                },
            },
        },
        {
            name: "Quantum-Safe Proposal Flow",
            commands: []BCTestCommand{
                {
                    name: "Prepare Quantum Proposal",
                    exec: func(t *testing.T, client types.ABCIClient) {
                        req := types.RequestPrepareProposal{
                            Txs: [][]byte{[]byte("quantum-safe-tx")},
                            MaxTxBytes: 1024,
                        }
                        res, err := client.PrepareProposalSync(req)
                        require.NoError(t, err, "proposal preparation failed")
                        assert.NotEmpty(t, res.Txs, "empty proposal response")
                    },
                },
                {
                    name: "Process Quantum Proposal",
                    exec: func(t *testing.T, client types.ABCIClient) {
                        req := types.RequestProcessProposal{
                            Txs: [][]byte{[]byte("quantum-safe-tx")},
                        }
                        res, err := client.ProcessProposalSync(req)
                        require.NoError(t, err, "proposal processing failed")
                        assert.Equal(t, types.ResponseProcessProposal_ACCEPT, res.Status)
                    },
                },
            },
        },
        {
            name: "AI-Optimized Transaction Flow",
            commands: []BCTestCommand{
                {
                    name: "Initial State Commit",
                    exec: func(t *testing.T, client types.ABCIClient) {
                        res, err := client.CommitSync()
                        require.NoError(t, err, "commit failed")
                        assert.NotNil(t, res.Data, "empty commit hash")
                    },
                },
                {
                    name: "AI-Routed Transaction",
                    exec: func(t *testing.T, client types.ABCIClient) {
                        req := types.RequestDeliverTx{
                            Tx: []byte("ai-optimized-tx"),
                        }
                        res, err := client.DeliverTxSync(req)
                        require.NoError(t, err, "transaction delivery failed")
                        assert.Equal(t, uint32(0), res.Code, "transaction rejected")
                    },
                },
                {
                    name: "Quantum-Safe Query",
                    exec: func(t *testing.T, client types.ABCIClient) {
                        req := types.RequestQuery{
                            Data: []byte("ai-optimized-tx"),
                            Path: "/store",
                            Prove: true,
                        }
                        res, err := client.QuerySync(req)
                        require.NoError(t, err, "query failed")
                        assert.Equal(t, uint32(0), res.Code, "query rejected")
                    },
                },
            },
        },
        {
            name: "Secure Key-Value Operations",
            commands: []BCTestCommand{
                {
                    name: "Set Quantum-Safe KV Pair",
                    exec: func(t *testing.T, client types.ABCIClient) {
                        req := types.RequestDeliverTx{
                            Tx: []byte("secure-key=secure-value"),
                        }
                        res, err := client.DeliverTxSync(req)
                        require.NoError(t, err, "kv set failed")
                        assert.Equal(t, uint32(0), res.Code, "kv set rejected")
                    },
                },
                {
                    name: "Verify Secure KV",
                    exec: func(t *testing.T, client types.ABCIClient) {
                        req := types.RequestQuery{
                            Data: []byte("secure-key"),
                            Path: "/store",
                            Prove: true,
                        }
                        res, err := client.QuerySync(req)
                        require.NoError(t, err, "kv verification failed")
                        assert.Equal(t, []byte("secure-value"), res.Value)
                    },
                },
            ],
        },
        // Add more Baron Chain specific test sequences here
    }

    for _, seq := range testSequences {
        t.Run(seq.name, func(t *testing.T) {
            for _, cmd := range seq.commands {
                t.Run(cmd.name, func(t *testing.T) {
                    select {
                    case <-ctx.Done():
                        t.Fatal("Baron Chain test sequence timeout")
                    default:
                        cmd.exec(t, env.client)
                    }
                })
            }
        })
    }
}

// Helper for hex debugging
func dumpHexData(data []byte) string {
    return hex.EncodeToString(data)
}
