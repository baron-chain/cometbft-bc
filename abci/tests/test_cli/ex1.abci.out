# Echo Command
echo hello
>> code: OK
>> data: hello
>> data.hex: 68656C6C6F

# Initial Info
info 
>> code: OK
>> data: {"size":0}
>> data.hex: 7B2273697A65223A307D

# First Transaction Block
prepare_proposal "abc"
>> code: OK
>> log: Succeeded. Tx: abc

process_proposal "abc"
>> code: OK
>> status: ACCEPT

commit 
>> code: OK
>> data.hex: 0000000000000000

# Transaction Delivery
deliver_tx "abc"
>> code: OK

# Info After First Transaction
info 
>> code: OK
>> data: {"size":1}
>> data.hex: 7B2273697A65223A317D

# Second Commit
commit 
>> code: OK
>> data.hex: 0200000000000000

# Query First Transaction
query "abc"
>> code: OK
>> log: exists
>> height: 2
>> key: abc
>> key.hex: 616263
>> value: abc
>> value.hex: 616263

# Key-Value Transaction
deliver_tx "def=xyz"
>> code: OK

commit 
>> code: OK
>> data.hex: 0400000000000000

# Query Key-Value
query "def"
>> code: OK
>> log: exists
>> height: 3
>> key: def
>> key.hex: 646566
>> value: xyz
>> value.hex: 78797A

# Proposal Handling
prepare_proposal "preparedef"
>> code: OK
>> log: Succeeded. Tx: replacedef

process_proposal "replacedef"
>> code: OK
>> status: ACCEPT

process_proposal "preparedef"
>> code: OK
>> status: REJECT

# Empty Proposals
prepare_proposal 

process_proposal 
>> code: OK
>> status: ACCEPT

# Final Commit
commit 
>> code: OK
>> data.hex: 0400000000000000
